File: flutter/lib/src/components/symbol_container.dart
import 'package:lemonade_design_system/lemonade_design_system.dart';

/// Voice variants available for [LemonadeSymbolContainer].
enum LemonadeSymbolContainerVoice {
  /// Neutral voice - default styling
  neutral,

  /// Critical voice - error/danger styling
  critical,

  /// Warning voice - caution styling
  warning,

  /// Info voice - informational styling
  info,

  /// Positive voice - success styling
  positive,

  /// Brand voice - brand color styling
  brand,

  /// Brand subtle voice - subtle brand styling
  brandSubtle,
}

/// Size variants available for [LemonadeSymbolContainer].
enum LemonadeSymbolContainerSize {
  /// Extra small size
  xSmall,

  /// Small size
  small,

  /// Medium size
  medium,

  /// Large size
  large,

  /// Extra large size
  xLarge,
}

/// {@template LemonadeSymbolContainer}
/// A symbol container widget from the Lemonade Design System.
///
/// A [LemonadeSymbolContainer] is a versatile container used to display
/// an icon, text, or custom content with consistent sizing and styling.
///
/// ## Example with icon
/// ```dart
/// LemonadeSymbolContainer.icon(
///   icon: LemonadeIcons.heart,
///   voice: LemonadeSymbolContainerVoice.info,
///   size: LemonadeSymbolContainerSize.medium,
/// )
/// ```
///
/// ## Example with text
/// ```dart
/// LemonadeSymbolContainer.text(
///   text: 'A',
///   voice: LemonadeSymbolContainerVoice.positive,
/// )
/// ```
///
/// See also:
/// - [LemonadeSymbolContainerTheme], for theme configuration
/// {@endtemplate}
class LemonadeSymbolContainer extends StatelessWidget {
  /// Creates a symbol container with an icon.
  const LemonadeSymbolContainer.icon({
    required this.icon,
    this.voice = LemonadeSymbolContainerVoice.neutral,
    this.size = LemonadeSymbolContainerSize.medium,
    this.semanticIdentifier,
    this.semanticLabel,
    super.key,
  }) : text = null,
       child = null;

  /// Creates a symbol container with text.
  const LemonadeSymbolContainer.text({
    required String this.text,
    this.voice = LemonadeSymbolContainerVoice.neutral,
    this.size = LemonadeSymbolContainerSize.medium,
    this.semanticIdentifier,
    this.semanticLabel,
    super.key,
  }) : icon = null,
       child = null;

  /// Creates a symbol container with custom content.
  const LemonadeSymbolContainer.custom({
    required Widget this.child,
    this.voice = LemonadeSymbolContainerVoice.neutral,
    this.size = LemonadeSymbolContainerSize.medium,
    this.semanticIdentifier,
    this.semanticLabel,
    super.key,
  }) : icon = null,
       text = null;

  /// The icon to display (if using icon variant).
  final LemonadeIcons? icon;

  /// The text to display (if using text variant).
  final String? text;

  /// Custom content (if using custom variant).
  final Widget? child;

  /// The voice/tone of the container. Defaults to [LemonadeSymbolContainerVoice.neutral].
  final LemonadeSymbolContainerVoice voice;

  /// The size of the container.
  /// Defaults to [LemonadeSymbolContainerSize.medium].
  final LemonadeSymbolContainerSize size;

  /// An identifier for the container for accessibility and testing.
  final String? semanticIdentifier;

  /// A label for the container for accessibility purposes.
  final String? semanticLabel;

  Color _getTintColor(LemonadeSemanticColors colors) {
    return switch (voice) {
      LemonadeSymbolContainerVoice.neutral => colors.content.contentPrimary,
      LemonadeSymbolContainerVoice.critical => colors.content.contentCritical,
      LemonadeSymbolContainerVoice.warning => colors.content.contentCaution,
      LemonadeSymbolContainerVoice.info => colors.content.contentInfo,
      LemonadeSymbolContainerVoice.positive => colors.content.contentPositive,
      LemonadeSymbolContainerVoice.brand => colors.content.contentOnBrandHigh,
      LemonadeSymbolContainerVoice.brandSubtle =>
        colors.content.contentOnBrandHigh,
    };
  }

  Color _getContainerColor(LemonadeSemanticColors colors) {
    return switch (voice) {
      LemonadeSymbolContainerVoice.neutral => colors.background.bgNeutralSubtle,
      LemonadeSymbolContainerVoice.critical =>
        colors.background.bgCriticalSubtle,
      LemonadeSymbolContainerVoice.warning => colors.background.bgCautionSubtle,
      LemonadeSymbolContainerVoice.info => colors.background.bgInfoSubtle,
      LemonadeSymbolContainerVoice.positive =>
        colors.background.bgPositiveSubtle,
      LemonadeSymbolContainerVoice.brand => colors.background.bgBrand,
      LemonadeSymbolContainerVoice.brandSubtle =>
        colors.background.bgBrandSubtle,
    };
  }

  _SymbolContainerDimensions _getDimensions(
    LemonadeSymbolContainerTheme theme,
  ) {
    return switch (size) {
      LemonadeSymbolContainerSize.xSmall => _SymbolContainerDimensions(
        containerSize: theme.xSmallContainerSize,
        iconSize: LemonadeIconSize.xSmall,
      ),
      LemonadeSymbolContainerSize.small => _SymbolContainerDimensions(
        containerSize: theme.smallContainerSize,
        iconSize: LemonadeIconSize.small,
      ),
      LemonadeSymbolContainerSize.medium => _SymbolContainerDimensions(
        containerSize: theme.mediumContainerSize,
        iconSize: LemonadeIconSize.medium,
      ),
      LemonadeSymbolContainerSize.large => _SymbolContainerDimensions(
        containerSize: theme.largeContainerSize,
        iconSize: LemonadeIconSize.large,
      ),
      LemonadeSymbolContainerSize.xLarge => _SymbolContainerDimensions(
        containerSize: theme.xLargeContainerSize,
        iconSize: LemonadeIconSize.xLarge,
      ),
    };
  }

  TextStyle _getTextStyle(
    LemonadeTypography typography,
    LemonadeSymbolContainerSize size,
  ) {
    return switch (size) {
      LemonadeSymbolContainerSize.xSmall => typography.bodyXSmallSemibold,
      LemonadeSymbolContainerSize.small => typography.bodySmallSemibold,
      LemonadeSymbolContainerSize.medium => typography.bodySmallSemibold,
      LemonadeSymbolContainerSize.large => typography.bodyLargeSemibold,
      LemonadeSymbolContainerSize.xLarge => typography.bodyXLargeSemibold,
    };
  }

  @override
  Widget build(BuildContext context) {
    final theme = LemonadeTheme.of(context);
    final symbolContainerTheme = theme.components.symbolContainerTheme;
    final colors = theme.colors;
    final typography = theme.typography;

    final tintColor = _getTintColor(colors);
    final containerColor = _getContainerColor(colors);
    final dimensions = _getDimensions(symbolContainerTheme);

    Widget content;
    if (icon != null) {
      content = LemonadeIcon(
        icon: icon!,
        size: dimensions.iconSize,
        color: tintColor,
      );
    } else if (text != null) {
      content = Text(
        text!,
        style: _getTextStyle(typography, size).copyWith(color: tintColor),
      );
    } else {
      content = child!;
    }

    return Semantics(
      label: semanticLabel ?? text,
      identifier: semanticIdentifier,
      child: Container(
        width: dimensions.containerSize,
        height: dimensions.containerSize,
        decoration: BoxDecoration(
          color: containerColor,
          shape: BoxShape.circle,
        ),
        child: Center(child: content),
      ),
    );
  }
}

class _SymbolContainerDimensions {
  const _SymbolContainerDimensions({
    required this.containerSize,
    required this.iconSize,
  });

  final double containerSize;
  final LemonadeIconSize iconSize;
}

File: kmp/ui/src/commonMain/kotlin/com/teya/lemonade/SymbolContainer.kt
package com.teya.lemonade

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.requiredSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import com.teya.lemonade.core.LemonadeAssetSize
import com.teya.lemonade.core.LemonadeIcons
import com.teya.lemonade.core.LemonadeTextStyle
import com.teya.lemonade.core.SymbolContainerSize
import com.teya.lemonade.core.SymbolContainerVoice
import org.jetbrains.compose.ui.tooling.preview.PreviewParameter
import org.jetbrains.compose.ui.tooling.preview.PreviewParameterProvider

/**
 * A versatile container used to display an icon, brand logo, or image.
 *  Supports consistent sizing and different tone of voice.
 * ## Usage
 * ```kotlin
 * LemonadeUi.SymbolContainer(
 *     icon = LemonadeIcons.Heart,
 *     contentDescription = "Content Description",
 *     voice = SymbolContainerVoice.Info,
 *     size = SymbolContainerSize.Small,
 * )
 * ```
 * @param icon - [LemonadeIcons] to be displayed inside the container.
 * @param contentDescription - the **localized** content description for the [icon].
 * @param modifier - Optional, the [Modifier] to be applied to the base component.
 * @param voice - [SymbolContainerVoice] to define the tone of voice. This will effectively define
 *  color of the background and the tint for the [icon]. Defaults to [SymbolContainerVoice.Neutral].
 * @param size - [SymbolContainerSize] to define the container's size. Defaults to [SymbolContainerSize.Medium].
 */
@Composable
public fun LemonadeUi.SymbolContainer(
    icon: LemonadeIcons,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    voice: SymbolContainerVoice = SymbolContainerVoice.Neutral,
    size: SymbolContainerSize = SymbolContainerSize.Medium,
) {
    CoreSymbolContainer(
        voice = voice,
        size = size,
        modifier = modifier,
        contentSlot = {
            LemonadeUi.Icon(
                icon = icon,
                size = LocalSymbolContainerPlatformDimensions.current.lemonadeIconSize,
                contentDescription = contentDescription,
                tint = voice.tintColor,
            )
        },
    )
}

/**
 * A versatile container used to display an icon, brand logo, or image.
 *  Supports consistent sizing and different tone of voice.
 * ## Usage
 * ```kotlin
 * LemonadeUi.SymbolContainer(
 *     text = "W",
 *     voice = SymbolContainerVoice.Info,
 *     size = SymbolContainerSize.Small,
 * )
 * ```
 * @param text - [Text] to be displayed inside the container.
 * @param modifier - Optional, the [Modifier] to be applied to the base component.
 * @param voice - [SymbolContainerVoice] to define the tone of voice. This will effectively define
 *  color of the background and the color for the [text]. Defaults to [SymbolContainerVoice.Neutral].
 * @param size - [SymbolContainerSize] to define the container's size. Defaults to [SymbolContainerSize.Medium].
 */
@Composable
public fun LemonadeUi.SymbolContainer(
    text: String,
    modifier: Modifier = Modifier,
    voice: SymbolContainerVoice = SymbolContainerVoice.Neutral,
    size: SymbolContainerSize = SymbolContainerSize.Medium,
) {
    CoreSymbolContainer(
        voice = voice,
        size = size,
        modifier = modifier,
        contentSlot = {
            LemonadeUi.Text(
                text = text,
                color = voice.tintColor,
                textStyle = LocalSymbolContainerPlatformDimensions.current.textStyle,
            )
        },
    )
}

/**
 * A versatile container used to display an icon, brand logo, or image.
 *  Supports consistent sizing and different tone of voice.
 * ## Usage
 * ```kotlin
 * LemonadeUi.SymbolContainer(
 *     voice = SymbolContainerVoice.Info,
 *     size = SymbolContainerSize.Small,
 *     contentSlot = {
 *          Image(
 *              painter = ...,
 *              ...
 *          )
 *     },
 * )
 * ```
 * @param contentSlot - A Composable slot for generic content to be added as the content of the container.
 * @param modifier - Optional, the [Modifier] to be applied to the base component.
 * @param voice - [SymbolContainerVoice] to define the tone of voice. This will effectively define
 *  color of the background. Defaults to [SymbolContainerVoice.Neutral].
 * @param size - [SymbolContainerSize] to define the container's size. Defaults to [SymbolContainerSize.Medium].
 */
@Composable
public fun LemonadeUi.SymbolContainer(
    contentSlot: @Composable BoxScope.() -> Unit,
    modifier: Modifier = Modifier,
    voice: SymbolContainerVoice = SymbolContainerVoice.Neutral,
    size: SymbolContainerSize = SymbolContainerSize.Medium,
) {
    CoreSymbolContainer(
        voice = voice,
        size = size,
        modifier = modifier,
        contentSlot = {
            Box(
                content = contentSlot,
                contentAlignment = Alignment.Center,
                modifier = Modifier.requiredSize(
                    size = LocalSymbolContainerPlatformDimensions.current.contentSize,
                ),
            )
        },
    )
}

@Composable
private fun CoreSymbolContainer(
    contentSlot: @Composable BoxScope.() -> Unit,
    voice: SymbolContainerVoice,
    size: SymbolContainerSize,
    modifier: Modifier = Modifier,
) {
    val dimensions = size.defaultSymbolContainerPlatformDimensions()
    CompositionLocalProvider(LocalSymbolContainerPlatformDimensions provides dimensions) {
        Box(
            content = contentSlot,
            contentAlignment = Alignment.Center,
            modifier = modifier
                .clip(shape = LocalShapes.current.radiusFull)
                .background(color = voice.containerColor)
                .requiredSize(size = dimensions.containerSize),
        )
    }
}

private val LocalSymbolContainerPlatformDimensions =
    staticCompositionLocalOf<SymbolContainerPlatformDimensions> {
        error("Local Symbol container platform dimensions not initialized")
    }

private val SymbolContainerVoice.tintColor: Color
    @Composable get() {
        return when (this) {
            SymbolContainerVoice.Neutral -> LocalColors.current.content.contentPrimary
            SymbolContainerVoice.Critical -> LocalColors.current.content.contentCritical
            SymbolContainerVoice.Warning -> LocalColors.current.content.contentCaution
            SymbolContainerVoice.Info -> LocalColors.current.content.contentInfo
            SymbolContainerVoice.Positive -> LocalColors.current.content.contentPositive
            SymbolContainerVoice.Brand -> LocalColors.current.content.contentOnBrandHigh
            SymbolContainerVoice.BrandSubtle -> LocalColors.current.content.contentOnBrandHigh
        }
    }

private val SymbolContainerVoice.containerColor: Color
    @Composable get() {
        return when (this) {
            SymbolContainerVoice.Neutral -> LocalColors.current.background.bgNeutralSubtle
            SymbolContainerVoice.Critical -> LocalColors.current.background.bgCriticalSubtle
            SymbolContainerVoice.Warning -> LocalColors.current.background.bgCautionSubtle
            SymbolContainerVoice.Info -> LocalColors.current.background.bgInfoSubtle
            SymbolContainerVoice.Positive -> LocalColors.current.background.bgPositiveSubtle
            SymbolContainerVoice.Brand -> LocalColors.current.background.bgBrand
            SymbolContainerVoice.BrandSubtle -> LocalColors.current.background.bgBrandSubtle
        }
    }

private data class SymbolContainerPlatformDimensions(
    val containerSize: Dp,
    val contentSize: Dp,
    val lemonadeIconSize: LemonadeAssetSize,
    val textStyle: LemonadeTextStyle,
)

@Composable
private fun SymbolContainerSize.defaultSymbolContainerPlatformDimensions(): SymbolContainerPlatformDimensions =
    when (this) {
        SymbolContainerSize.XSmall -> SymbolContainerPlatformDimensions(
            containerSize = LocalSizes.current.size600,
            contentSize = LocalSizes.current.size300,
            lemonadeIconSize = LemonadeAssetSize.XSmall,
            textStyle = LocalTypographies.current.bodyXSmallSemiBold,
        )

        SymbolContainerSize.Small -> SymbolContainerPlatformDimensions(
            containerSize = LocalSizes.current.size800,
            contentSize = LocalSizes.current.size400,
            lemonadeIconSize = LemonadeAssetSize.Small,
            textStyle = LocalTypographies.current.bodySmallSemiBold,
        )

        SymbolContainerSize.Medium -> SymbolContainerPlatformDimensions(
            containerSize = LocalSizes.current.size1000,
            contentSize = LocalSizes.current.size500,
            lemonadeIconSize = LemonadeAssetSize.Medium,
            textStyle = LocalTypographies.current.bodySmallSemiBold,
        )

        SymbolContainerSize.Large -> SymbolContainerPlatformDimensions(
            containerSize = LocalSizes.current.size1200,
            contentSize = LocalSizes.current.size600,
            lemonadeIconSize = LemonadeAssetSize.Large,
            textStyle = LocalTypographies.current.bodyLargeSemiBold,
        )

        SymbolContainerSize.XLarge -> SymbolContainerPlatformDimensions(
            containerSize = LocalSizes.current.size1600,
            contentSize = LocalSizes.current.size800,
            lemonadeIconSize = LemonadeAssetSize.XLarge,
            textStyle = LocalTypographies.current.bodyXLargeSemiBold,
        )
    }

private data class SymbolContainerPreviewData(
    val content: Any,
    val size: SymbolContainerSize,
    val voice: SymbolContainerVoice,
)

private class SymbolContainerPreviewProvider :
    PreviewParameterProvider<SymbolContainerPreviewData> {
    override val values: Sequence<SymbolContainerPreviewData> = buildAllVariants()

    private fun buildAllVariants(): Sequence<SymbolContainerPreviewData> =
        buildList {
            SymbolContainerVoice.entries.forEach { voice ->
                listOf("A", LemonadeIcons.Heart).forEach { content ->
                    SymbolContainerSize.entries.forEach { size ->
                        add(
                            SymbolContainerPreviewData(
                                content = content,
                                size = size,
                                voice = voice,
                            ),
                        )
                    }
                }
            }
        }.asSequence()
}

@Suppress("UnusedPrivateMember")
@LemonadePreview
@Composable
private fun SymbolContainerPreview(
    @PreviewParameter(SymbolContainerPreviewProvider::class)
    previewData: SymbolContainerPreviewData,
) {
    when (previewData.content) {
        is LemonadeIcons -> {
            LemonadeUi.SymbolContainer(
                icon = previewData.content,
                size = previewData.size,
                voice = previewData.voice,
                contentDescription = "Content Description",
            )
        }

        is String -> {
            LemonadeUi.SymbolContainer(
                text = previewData.content,
                size = previewData.size,
                voice = previewData.voice,
            )
        }
    }
}
